### ISSUE TITLE
Diagnostic recovery can suppress invalid specialization diagnostics

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#4299]
The following snippet should produce an error:

```py
from typing import Any, Callable

def infer[T: int](value: T, upper: Callable[[T], None]) -> T:
    return value

def example(value: Any, upper: Callable[[bytes], None]):
    reveal_type(infer(value, upper))  # revealed: Any
```

It seems that we recognize that the upper bounds are incompatible and [return an error from the solver](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/constraints.rs#L4004). However, the fallback specialization is `Any`, which loses the fact that the inferred upper bound `T :> bytes` did not satisfy the declared upper bound `T: int`.
